src/ConfigFileParser.cpp

```diff
@@ -34,83 +36,122 @@ ConfigFileParser::ConfigFileParser(void)
-
-			if (bracketIndex != std::string::npos) {
-				std::string withoutBracket = m_currentLine.substr(bracketIndex + 1, m_currentLine.length());
-				size_t firstNotWhiteSpaceIndex = withoutBracket.find_first_not_of(whitespace);
-				if (firstNotWhiteSpaceIndex == std::string::npos) {
-					readAndTrimLine();
-				} else
-					m_currentLine = withoutBracket.substr(firstNotWhiteSpaceIndex, withoutBracket.length());
-			} else
-				readAndTrimLine();
-
-			while (m_currentLine != "}") {
-				readServerConfigLine();
-				if (m_currentLine == "}")
-					break;
-				readAndTrimLine();
-			}
-			m_serverIndex++;
-		} else if (getDirective(m_currentLine).empty())
-			continue;
-		else
+	if (!isValidBlockBeginn(HttpBlock))
+		return m_configFile;
+
+	skipBlockBegin(HttpBlock);
+
+	while (m_configFileContent[m_configFileIndex] != '}') {
+		if (isValidBlockBeginn(ServerBlock))
+			readServerBlock();
+		else if (std::isspace(m_configFileContent[m_configFileIndex]) == 0) {
 			throw std::runtime_error("Invalid directive");
+		}
+		m_configFileIndex++;
+	}
+
+	for (std::vector<ServerBlockConfig>::const_iterator serverIt = m_serverBlocksConfig.begin();
+		 serverIt != m_serverBlocksConfig.end(); serverIt++) {
+		processServerContent(*serverIt);
+		m_serverIndex++;
 	}
-	if (m_configFile.servers.empty())
-		throw std::runtime_error("No server(s) in config file");
 
 	return m_configFile;
 }
 
+/*********************/
+/* CHECKER FUNCTIONS */
+/*********************/
+
 /**
- * @brief Checks if there are no open brackets
+ * @brief Looks for a keyword in the m_configFileContent (eg. http, server, location)
  *
- * If a opening bracket is found, it is pushed onto the brackets stack
+ * @param keyword The keyword to look for
+ * @return true If the keyword is found
+ * @return false If the keyword is not found
+ */
+bool ConfigFileParser::isKeyword(const std::string& keyword, size_t startIndex) const
+{
+	std::string string = m_configFileContent.substr(startIndex, keyword.length());
+	return string == keyword;
+}
+
+/**
+ * @brief Checks if the beginning of a block is valid
+ *
+ * @param block The block to check
+ * @return true If the beginning of the block is valid
+ * @return false If the beginning of the block is not valid
+ */
+bool ConfigFileParser::isValidBlockBeginn(Block block)
+{
+	size_t index = m_configFileIndex;
+
+	while (std::isspace(m_configFileContent[index]) != 0)
+		index++;
+
+	if (block == HttpBlock) {
+		if (!isKeyword(convertBlockToString(HttpBlock), index))
+			return false;
+	} else if (block == ServerBlock) {
+		if (!isKeyword(convertBlockToString(ServerBlock), index))
+			return false;
+	} else if (block == LocationBlock) {
+		if (!isKeyword(convertBlockToString(LocationBlock), index))
+			return false;
+	}
+
+	index += convertBlockToString(block).length();
+
+	while (std::isspace(m_configFileContent[index]) != 0)
+		index++;
+
+	if (block == LocationBlock)
+		skipLocationBlockPath(index);
+
+	index++;
```

Comment:
Why are you increasing index here, and then decreasing it afterwards to check the file content?

Reply:
This was a remainder of my approach when the index var was not on the stack but on the heap.
I wanted to increment the index but also check the value before the incrementation. 

Removed the incrementation and the decrementation.